hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/TimeoutOptimizer.java

```diff
@@ -0,0 +1,227 @@
+        }
+        else if (opType == AbfsRestOperationType.GetAcl) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_GET_ACL_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.GetPathProperties) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_GET_PATH_PROPERTIES_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.SetPathProperties) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_PATH_PROPERTIES_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.SetAcl) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_ACL_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.SetOwner) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_OWNER_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.SetPermissions) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_SET_PERMISSIONS_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.Append) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_APPEND_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.CheckAccess) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_CHECK_ACCESS_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.GetPathStatus) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_GET_PATH_STATUS_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.Flush) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_FLUSH_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.ReadFile) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_READFILE_REQUEST_TIMEOUT);
+        }
+        else if (opType == AbfsRestOperationType.LeasePath) {
+            timeout = abfsConfiguration.get(ConfigurationKeys.AZURE_LEASE_PATH_REQUEST_TIMEOUT);
+        }
+        if (timeout == null) {
+            timeout = DEFAULT_TIMEOUT;
+        }
+        requestTimeout = Integer.parseInt(timeout);
+        readTimeout = requestTimeout;
+        connTimeout = requestTimeout - 1;
+        updateUrl();
+    }
+
+    private void updateTimeouts(int retryCount) {
+        if (retryCount == 0) {
+            return;
+        }
+        int maxRetryCount = retryPolicy.getRetryCount();
+        if (retryCount <= maxRetryCount && timeoutIncRate > 0) {
+            // retry count is still valid
+            // timeout increment rate is a valid value
+            if ((requestTimeout * timeoutIncRate) > maxReqTimeout) {
+                requestTimeout = maxReqTimeout;
+            } else {
+                requestTimeout *= timeoutIncRate;
+            }
+            readTimeout = requestTimeout;
+            connTimeout = requestTimeout - 1;
```

Comment:
We understand that the request can take maximum of (connTimeout + requestTimeout) and JDK will not throw any issue? Feel that connTimeout can be kept more aggressive.

Why is readTimeout, connTimeout a function of requestTimeout? Feel that diff types of exception should change diff parameters.

CC: @snvijaya @anmolanmol1234

Reply:
ReadTimeout can be a function of requestTimeout, but connectionTimeout is totally independent to requestTimeout. Until connection is not established, there is no significance of requestTimeout as server would not have started processing.

Reply:
Made the necessary changes.